x/epochs/keeper/abci.go

```diff
@@ -0,0 +1,104 @@
+package keeper
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/ExocoreNetwork/exocore/x/epochs/types"
+)
+
+// BeginBlocker is used to start or end the epochs, amongst the epochs currently
+// in the store.
+func (k Keeper) BeginBlocker(ctx sdk.Context) {
+	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), telemetry.MetricKeyBeginBlocker)
+	logger := k.Logger(ctx)
+	k.IterateEpochInfos(
+		ctx,
+		func(_ int64, epochInfo types.EpochInfo) (stop bool) {
+			if ctx.BlockTime().Before(epochInfo.StartTime) {
+				// short circuit if this epoch is not yet scheduled to start
+				return false
+			}
+			epochEndTime := epochInfo.CurrentEpochStartTime.Add(epochInfo.Duration)
+			// are we starting this identifier for the first time?
+			isFirstTick := !epochInfo.EpochCountingStarted
+			// is this the end of the current tick?
+			isTickEnding := ctx.BlockTime().After(epochEndTime)
+			// if either of those conditions are true, we will start an epoch
+			isEpochStart := isTickEnding || isFirstTick
+
+			if !isEpochStart {
+				return false
+			}
+			// if we reach here, we are starting a new epoch. this means, we set its height.
+			epochInfo.CurrentEpochStartHeight = ctx.BlockHeight()
+
+			if isFirstTick {
+				epochInfo.EpochCountingStarted = true
+				// even if the genesis file may start at a different number, we will reset to 1.
+				epochInfo.CurrentEpoch = 1
+				epochInfo.CurrentEpochStartTime = epochInfo.StartTime
+				// we don't call BeforeEpochStart here because it is the first epoch.
+			} else {
+				// if we are here, isTickEnding is true but isFirstTick is false.
+				// in other words, epoch i is ending and epoch i+1 is starting.
+				logger.Info(
+					"ending epoch",
+					"identifier", epochInfo.Identifier,
+					"number", epochInfo.CurrentEpoch,
+				)
+				ctx.EventManager().EmitEvent(
+					sdk.NewEvent(
+						types.EventTypeEpochEnd,
+						sdk.NewAttribute(
+							types.AttributeEpochIdentifier, epochInfo.Identifier,
+						),
+						sdk.NewAttribute(
+							types.AttributeEpochNumber,
+							strconv.FormatInt(
+								epochInfo.CurrentEpoch, 10,
+							),
+						),
+					),
+				)
+				// NOTE: this hook is called BEFORE the new epoch info is saved.
+				k.Hooks().AfterEpochEnd(ctx, epochInfo.Identifier, epochInfo.CurrentEpoch)
```

Comment:
In the previously imported implementation, the `AfterEpochEnd` function was called after incrementing the `epochInfo.CurrentEpoch`, resulting in the same epoch (bearing the increased number) ending and then starting immediately afterwards

Reply:
As a result of this bug, the effective opt-out completion epoch was reduced by 1. 

For example, with the number of unbonding epochs set to 2, if an operator opts out during epoch 5, they will stop validating at the end of epoch 5 and complete the opt out at the end of epoch 7. This is accurately scheduled with or without the bug being present in the `x/epochs` module, because the currently running epoch number is accurately fetched during the time of opt out.

With the bug, however, the epoch reported in the `AfterEpochEnd` hook is 7 even though the epoch currently ending is 6.